File: 2021_summer_algorithm_study_intermediate/using_python/15816.py
"""
input :
3
1 10 20
4
2 1 20
1 5
2 1 20
2 1 1

output:
17
16
0
"""
import sys

sys.stdin = open("input.txt", "r")
input = sys.stdin.readline

PIV = 1 << 22
tree = [0 for _ in range(2 * PIV)]


def update(x):
    global PIV
    x += PIV
    if tree[x]: return

    tree[x] += 1
    while True:
        x >>= 1
        if x == 0: return
        tree[x] += 1


def query(l, r):
    global PIV
    l += PIV;
    r += PIV;

    ret = 0
    while l <= r:
        if l & 1:
            ret += tree[l]
            l += 1
        if not (r & 1):
            ret += tree[r]
            r -= 1
        l >>= 1;
        r >>= 1;
    return ret


if __name__ == "__main__":
    N = int(input())
    init_updates = list(map(int, input().rstrip().split()))
    cmds = []
    nums = []
    for num in init_updates:
        nums.append(num)

    # query들을 미리 전처리
    M = int(input())
    for _ in range(M):
        line = list(map(int, input().rstrip().split()))
        cmds.append(line)
        if line[0] == 1:
            nums.append(line[1])
        else:
            nums.append(line[1])
            nums.append(line[2])

    # 좌표 압축
    nums = list(set(nums))
    nums.sort()
    Q2I = dict()
    for ind, q in enumerate(nums):
        Q2I[q] = ind

    # 미리 주어진 update 쿼리 처리
    for num in init_updates:
        update(Q2I[num])

    # cmd 시퀀스를 다시 하나씩 보면서, 업데이트 쿼리와 구간합 쿼리를 수행
    for cmd in cmds:
        if cmd[0] == 1:
            x = cmd[1]
            update(Q2I[x])
        elif cmd[0] == 2:
            l, r = cmd[1], cmd[2]

            tot = query(Q2I[l], Q2I[r])
            print(r - l + 1 - tot)

File: 2021_summer_algorithm_study_intermediate/using_cpp/15816.cpp
/*
input :
3
1 10 20
4
2 1 20
1 5
2 1 20
2 1 1

output :
17 16 0
*/

#include <iostream>
#include <algorithm>
#include <vector>
#include <map>
#define PIV (1 << 22)
using namespace std;
using ll = long long;

// 전역 변수 선언
ll tree[2 * PIV];
ll N, M;
vector<ll> NUMS;

ll init_updates[1000003];
ll cmd[1000003][3];


void update(ll x){
    x += PIV;
    if (tree[x]) return;

    tree[x]++;
    while (x >>= 1){
        tree[x]++;
    }
}

ll query(ll l, ll r){
    l += PIV, r += PIV;

    ll ret = 0;
    while(l <= r){
        if (l & 1) ret += tree[l++];
        if (!(r & 1)) ret += tree[r--];
        l >>= 1, r >>= 1;
    }
    return ret;
}


int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    freopen("input.txt", "r", stdin);

    cin >> N;
    for (int i = 0; i < N; i++){
        cin >> init_updates[i];
        NUMS.push_back(init_updates[i]);
    }

    int type;
    cin >> M;
    for (int i = 0; i < M; i++){
        cin >> type;
        cmd[i][0] = type;
        switch (type){
            case 1:
                cin >> cmd[i][1];
                NUMS.push_back(cmd[i][1]);
                break;
            case 2:
                cin >> cmd[i][1] >> cmd[i][2];
                NUMS.push_back(cmd[i][1]);
                NUMS.push_back(cmd[i][2]);
                break;
            default:
                // no action
                break;
        }
    }

    // 좌표압축
    sort(NUMS.begin(), NUMS.end());
    NUMS.resize(unique(NUMS.begin(), NUMS.end()) - NUMS.begin());

    // init update 쿼리들 처리
    for (int i = 0; i < N; i++){
        int k = lower_bound(NUMS.begin(), NUMS.end(), init_updates[i]) - NUMS.begin();
        update(k);
    }

    // after update 및 구간합 쿼리들 처리
    for (int i = 0; i < M; i++){
        ll x, l, r; int comp_l, comp_r;
        switch (cmd[i][0]){
            case 1:
                x = lower_bound(NUMS.begin(), NUMS.end(), cmd[i][1]) - NUMS.begin();
                update(x);
                break;
            case 2:
                l = cmd[i][1], r = cmd[i][2];
                comp_l = lower_bound(NUMS.begin(), NUMS.end(), cmd[i][1]) - NUMS.begin();
                comp_r = lower_bound(NUMS.begin(), NUMS.end(), cmd[i][2]) - NUMS.begin();
                cout << r - l + 1 - query(comp_l, comp_r) << '\n';
                break;
            default:
                // no action
                break;
        }
    }
    return 0;
}